crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -402,88 +387,46 @@ impl TransactionBlock {
-    /// results and transaction blocks in `other`'s results. This may not be possible if the
-    /// resulting filter is inconsistent in some way (e.g. a filter that requires one field to be
-    /// two different values simultaneously).
-    pub(crate) fn intersect(self, other: Self) -> Option<Self> {
-        macro_rules! intersect {
-            ($field:ident, $body:expr) => {
-                intersect::field(self.$field, other.$field, $body)
-            };
-        }
-
-        Some(Self {
-            function: intersect!(function, FqNameFilter::intersect)?,
-            kind: intersect!(kind, intersect::by_eq)?,
-
-            after_checkpoint: intersect!(after_checkpoint, intersect::by_max)?,
-            at_checkpoint: intersect!(at_checkpoint, intersect::by_eq)?,
-            before_checkpoint: intersect!(before_checkpoint, intersect::by_min)?,
-
-            sign_address: intersect!(sign_address, intersect::by_eq)?,
-            recv_address: intersect!(recv_address, intersect::by_eq)?,
-            input_object: intersect!(input_object, intersect::by_eq)?,
-            changed_object: intersect!(changed_object, intersect::by_eq)?,
-
-            transaction_ids: intersect!(transaction_ids, |a, b| {
-                let a = BTreeSet::from_iter(a.into_iter());
-                let b = BTreeSet::from_iter(b.into_iter());
-                Some(a.intersection(&b).cloned().collect())
-            })?,
-        })
-    }
-}
-
-impl Paginated<Cursor> for StoredTransaction {
-    type Source = transactions::table;
-
-    fn filter_ge<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query
-            .filter(transactions::dsl::tx_sequence_number.ge(cursor.tx_sequence_number as i64))
-            .filter(
-                transactions::dsl::checkpoint_sequence_number
-                    .ge(cursor.tx_checkpoint_number as i64),
-            )
-    }
-
-    fn filter_le<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query
-            .filter(transactions::dsl::tx_sequence_number.le(cursor.tx_sequence_number as i64))
-            .filter(
-                transactions::dsl::checkpoint_sequence_number
-                    .le(cursor.tx_checkpoint_number as i64),
-            )
-    }
+        let (scan_prev_page, scan_lo) = tx_bounds.scan_prev_page_and_lo();
+        let (scan_next_page, scan_hi) = tx_bounds.scan_next_page_and_hi();
+
+        // When a `scan_limit` is specified, as long as we are still within the scanning range,
+        // there will be previous or subsequent pages to continue scanning through.
+        if scan_limit.is_some() {
+            if !conn.has_next_page && scan_next_page {
+                conn.has_next_page = true;
+                // Defer to the cursor of the last transaction scanned in the range, unless the
+                // result is empty in which case defer to the `scan_hi` value.
+                if conn.edges.is_empty() {
```

Comment:
Not sure why we need this check -- if there is an extra page to scan, then we should always override the cursor, even if edges is non-empty, to avoid rescanning the tail of transactions that we know aren't going to contain any results, between the last actual result and the end of the current scanning range.

Reply:
I don't think that's necessarily true. Say I have a scan limit of 100, and within that scan limited range, I've hit the first 50 limit. We should use the last element of the edges as the cursor, instead of overriding it, otherwise we'll skip out the next 50 of the scan limited range

Reply:
I see, yes that makes sense. In other words, if the result set is less than the specified limit, including an empty set, we know we have exhausted the candidates to scan, and so we can set the `endCursor` to that last transaction scanned rather than the last element in the result set

Reply:
Yes, I think that's right, and I think we also covered this part in our conversation yesterday, but we also need to check for and respect the case where we returned 25 results because there are literally only 25 results (this can happen if we reached the end of the scan range, or we hit a `before` cursor while paginating forward, for example).

Reply:
I might be losing the plot here :painlaugh:

If there are only 25 results within the total range to scan, we wouldn't know for sure until we've finished scanning through the total range right? So once we return the 25 results, we'll keep yielding empty results whose cursors are scan limited


If we hit a before cursor then we will prune it, and paginate query will always return hasNextPage = true

Reply:
> I might be losing the plot here :painlaugh:

...you and me both. But in this case, I think you haven't, your last sentence is what I'm referring to:

> If we hit a before cursor then we will prune it, and paginate query will always return hasNextPage = true

I'm saying that in this case we *shouldn't* update the cursor to point to the end of the scanned range, because the scanned range went past the end of the range defined by cursors. This isn't new though, this is what we talked about on Wednesday:

> - If paginate_results indicates that there is a next or previous page, then that is always correct (no false positives).
> - When paginating forward, if it indicates that there is no next page and we have a scan limit, this might be a false end, because we haven't scanned all transactions.
>   - Then we indicate there is a next page if the scan limit does not reach the end of the tx bounds.
>   - And we set the end cursor to the last transaction scanned.
> - Again, when paginating forward, if it indicates there is no previous page, and we are paginating from a scan limited cursor, this might be a false start, because paginate results expects to find the cursor in the results, and it couldn't (because it is scan limited).
>   - We indicate there is a previous page if the scan limited cursor does not reach the start of the tx bounds.
>   - We set the start cursor to one past the scan limited cursor (the first transaction scanned).

What we're talking about here is the first bullet (no false positives from `paginate_results`).